### PR TITLE
Add Stable Diffusion text to image workflow

### LINKED ISSUE #83: Add stable-diffusion as an optional flow executor for image generation
The stable diffusion model is not yet out, but as soon as it is released I will add a PR to allow it to be used in the list of executors for image generation. If I have time I will also add variation, embedding transitions (morphing gifs between embeddings), and inpainting.

### PR DESCRIPTION
**Pull request by AmericanPresidentJimmyCarter**
This pull requests adds support for Stable Diffusion. Please note that you will need access to a checkpoint for anything to work, in your `dalle/stable-diffusion/models/ldm/stable-diffusion-v1/` folder as `model.ckpt`. Although this model does not currently automatically download, it will upon release of the stable diffusion checkpoints to the general public.

Closes #83 .

If I have time before release I will add more features, like diffusion (image to image), inpainting, and others.

### CODE REVIEW

**Review by JoanFM — CHANGES_REQUESTED** [review]
The same idea why the last change was reverted. This makes the model very expensive to load. We need to find a way to make it optional or to have different flow yamls. Please @delgermurun help on defining this

File: executors/stable/executor.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+from ldm.util import instantiate_from_config
+from ldm.models.diffusion.ddim import DDIMSampler
+from ldm.models.diffusion.plms import PLMSSampler
+
+from jina import Executor, DocumentArray, Document, requests
+from omegaconf import OmegaConf
+
+
+class StableDiffusionConfig():
+    '''
```
Comment (JoanFM) [inline executors/stable/executor.py:25]: this looks like it should be a dataclass

File: executors/stable/executor.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    precision = 'autocast'
+
+
+def chunk(it, size):
+    it = iter(it)
+    return iter(lambda: tuple(islice(it, size)), ())
+
+
+def load_model_from_config(config, ckpt, verbose=False):
+    print(f"Loading model from {ckpt}")
```
Comment (JoanFM) [inline executors/stable/executor.py:51]: do not use print, use or pass a proper logger I would say

File: executors/stable/executor.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        print("unexpected keys:")
+        print(u)
+
+    model.cuda()
+    model.eval()
+    return model
+
+
+class StableDiffusionGenerator(Executor):
+    opt: StableDiffusionConfig = None
```
Comment (JoanFM) [inline executors/stable/executor.py:71]: what is the point if I cannot change it?

File: executors/stable/executor.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+    @requests(on='/')
+    def txt2img(self, docs: DocumentArray, parameters: Dict, **kwargs):
+        num_images = max(1, min(9, int(parameters.get('num_images', 1))))
+        request_time = time.time()
+        opt = self.opt
+        opt.n_samples = num_images
+        seed_everything(opt.seed)
+
+        device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
```
Comment (JoanFM) [inline executors/stable/executor.py:92]: this should be done at init time and not at every request

File: flow.yml [hunk 5]
```diff
@@ -36,7 +43,7 @@ executors:
       '/': rank
     tls: true
     external: true
-    needs: [dalle, diffusion]
+    needs: [stable, dalle, diffusion]
```
Comment (JoanFM) [inline flow.yml:46]: this does not make sense if u do not make stable need of gateway

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: executors/stable/executor.py (hunk 1 above)
Comment (AmericanPresidentJimmyCarter) [reply executors/stable/executor.py:25]: Yes, should be a subclass of `Executor`

Comment (AmericanPresidentJimmyCarter) [reply executors/stable/executor.py:25]: Sure

File: executors/stable/executor.py (hunk 2 above)
Comment (AmericanPresidentJimmyCarter) [reply executors/stable/executor.py:51]: Sure, sorry this was debug that got left in.

File: flow.yml (hunk 5 above)
Comment (AmericanPresidentJimmyCarter) [reply flow.yml:46]: Yeah, I wondered about this. I will remove it, thanks.

**Comment by AmericanPresidentJimmyCarter** [comment]
> The same idea why the last change was reverted. This makes the model very expensive to load. We need to find a way to make it optional or to have different flow yamls. Please @delgermurun help on defining this

@JoanFM So I figured out what was making it expensive and it presently is that:
1. Memory scales quadratically with the batch size (number of images).
2. Memory scales quadratically with the image resolution.

We can either make the image sizes smaller or alternatively do a hack when I generate in batches of 2 or 4 and then bump the seed in between them. It's not exactly ideal either way.

File: executors/stable/executor.py (hunk 4 above)
Comment (AmericanPresidentJimmyCarter) [reply executors/stable/executor.py:92]: Yeah, it was moved in latest (along with moving model init to ``__init__`) but I haven't pushed, working on img2img.

File: executors/stable/executor.py (hunk 3 above)
Comment (AmericanPresidentJimmyCarter) [reply executors/stable/executor.py:71]: I will default it to an instance and add ways to alter the keys via the `config.yml`.

**Comment by JoanFM** [comment]
> > The same idea why the last change was reverted. This makes the model very expensive to load. We need to find a way to make it optional or to have different flow yamls. Please @delgermurun help on defining this
> 
> @JoanFM So I figured out what was making it expensive and it presently is that:
> 
> 1. Memory scales quadratically with the batch size (number of images).
> 2. Memory scales quadratically with the image resolution.
> 
> We can either make the image sizes smaller or alternatively do a hack when I generate in batches of 2 or 4 and then bump the seed in between them. It's not exactly ideal either way.

@delgermurun please check how to do this? maybe we need to add options or what? Is it okey to add more layers of complexity? I think we should have different Flows with different options?

**Comment by AmericanPresidentJimmyCarter** [comment]
@JoanFM @delgermurun I thought about this last night and I think the way to go might be to add a virtual volume with a `flow.yml` that can be used instead of the default one in the docker. Since the `flow.yml` file contains all the settings more or less, people could tweak it to work with larger batches by making the images smaller for SD.

To address the "we only have so much memory issue", I would like to add an `enabled` flag to each executor in the `flow.yml`. When this flag is set to `false`, it will prevent the model from loading into the GPU and the gRPC calls from executing by returning an error. One potential problem I see is that this may cause complicated errors if you're disabling things which have many dependencies on other parts of the flow, but maybe in the future we could use a dependency graph to figure out if we accidentally broke something and alert the user. For now I'd like to leave dependency conflict resolution TODO and add a note in the repo.

It would be great if we had more simply the ability to hot load these models through the server, but I just can't see how we'd do that with the current state of pytorch GPU memory management. I haven't found a clear way to guarantee that all memory ends up getting freed when unloading models, and in the absence of that I think we will have a high likelihood of OOM-related crashes.

**Comment by AmericanPresidentJimmyCarter** [comment]
I have spent some time reading the docs. I believe that the easiest move is to selectively disable various flows is to simply rewrite the YAML as a python flow application and use environmental variables like `DISABLE_DALLE_MEGA` to selective prevent addition of a flow module/executor. This can be applied by flag when pulling and using the docker image, which would be a seamless experience for most users.

**Comment by AmericanPresidentJimmyCarter** [comment]
OK. `jina flow` doesn't allow you to use a python script, just YAML it looks, however, we can pass environmental variables there. I will figure something out.

File: flow.yml (hunk 5 above)
Comment (AmericanPresidentJimmyCarter) [reply flow.yml:46]: Actually, I get the error:
```
WARNI… Flow@77837 ['stable'] are "floating" in this flow     [08/19/22 14:43:56]
       with no deployment receiving from them, you may want                     
       to double check if it is intentional or some mistake
```

It does seem like it uses this gateway (`rerank`) to rank candidates returned from stable. I will leave it in place.

**Comment by AmericanPresidentJimmyCarter** [comment]
OK, a solution for disabling the executors is now in place with environmental variables.

Memory usage has been solved by allowing the user to specify which modules they want, either by using environmental variables (described in the README) or by specifying flags to the `flow.yml` tweaker `flow_parser.py`.

Memory usage for Stable Diffusion is now tweakable by altering the `n_iter`, `n_samples`, `height`, and `width` flags in `stable/config.yml`.

**Comment by AmericanPresidentJimmyCarter** [comment]
I started working on inpainting only to realize that it is an old latent inpainting model. When/if SD has a new one I can add it. Otherwise this might be ready to go.

**Comment by AmericanPresidentJimmyCarter** [comment]
I am still patching things up, so this is not quite ready for merge.

**Comment by AmericanPresidentJimmyCarter** [comment]
After some more testing I think it's good. Rebasing and ready for review.

**Comment by AmericanPresidentJimmyCarter** [comment]
I have been using this branch for about 72 hours now and I think it's more or less good to go as soon as stable diffusion's model can be auto-downloaded from huggingface, which may happen tomorrow.

**Comment by delgermurun** [comment]
@AmericanPresidentJimmyCarter Could you please push `flow_parser.py`?

**Comment by AmericanPresidentJimmyCarter** [comment]
@delgermurun My mistake, forgot to `git add`.

**Comment by AmericanPresidentJimmyCarter** [comment]
Use the weights from here.

https://huggingface.co/CompVis/stable-diffusion-v-1-4-original/blob/main/sd-v1-4.ckpt

**Comment by delgermurun** [comment]
@AmericanPresidentJimmyCarter 
Were weights published? The [link](https://huggingface.co/CompVis/stable-diffusion-v-1-4-original/blob/main/sd-v1-4.ckpt) shows a 404 page.

Please add the download command to Dockerfiles (both the main and executor docker files) if weights are released.

**Comment by delgermurun** [comment]
The additional 10GB of GPU VRAM from Stable Diffusion makes Dalle-Flow require 31GB GPU VRAM, which is too much.

Let's keep the old `flow.yml`, which will require less resources by default (21GB).
And then, let's add instructions for switching between image generators. I think the flow parser script will make it easy to enable/disable executors.

@AmericanPresidentJimmyCarter, what do you think?

**Comment by AmericanPresidentJimmyCarter** [comment]
> @AmericanPresidentJimmyCarter Were weights published? The [link](https://huggingface.co/CompVis/stable-diffusion-v-1-4-original/blob/main/sd-v1-4.ckpt) shows a 404 page.
> 
> Please add the download command to Dockerfiles (both the main and executor docker files) if weights are released.

@delgermurun That's the problem, I can't. The links 404 because you aren't logged into huggingface, and you are only allowed to download the SD model if you agree to their [terms and conditions](https://github.com/CompVis/stable-diffusion/blob/main/LICENSE).

I can point people towards the model in the README.md and give instructions to mount a virtual volume so that the model can be accessed via docker, but unfortunately to legally obtain the model requires the user to access huggingface and download it.

**Comment by AmericanPresidentJimmyCarter** [comment]
> The additional 10GB of GPU VRAM from Stable Diffusion makes Dalle-Flow require 31GB GPU VRAM, which is too much.
> 
> Let's keep the old `flow.yml`, which will require less resources by default (21GB). And then, let's add instructions for switching between image generators. I think the flow parser script will make it easy to enable/disable executors.
> 
> @AmericanPresidentJimmyCarter, what do you think?

That's fine. I can just change all the default commands to run with stable diffusion disabled, which is the previous default (glid3xl and dalle-mega), then add a specific section in the guide for how to setup and configure SD if people wish to enable it.

**Comment by delgermurun** [comment]
> I can point people towards the model in the README.md and give instructions to mount a virtual volume so that the model can be accessed via docker, but unfortunately to legally obtain the model requires the user to access huggingface and download it.

I see. Then we have no choice but to instruct users to download themselves and mount volumes to Docker themselves.

**Comment by AmericanPresidentJimmyCarter** [comment]
> I see. Then we have no choice but to instruct users to download themselves and mount volumes to Docker.

OK -- I will work on this later tonight when I have some free time and test it out. For `flow_parser.py` I will change the flag for SD to `--enable-stable-diffusion` and make the argumentless default with it disabled, then update the README with specific instructions and images for installation of SD.

**Comment by AmericanPresidentJimmyCarter** [comment]
@delgermurun Branch updated, k_lms sampler added (default in dreamstudio/huggingface diffusers), docker tested, new instructions added, should be good to go.

**Comment by delgermurun** [comment]
@AmericanPresidentJimmyCarter please remove (or comment out) `stable` executor from `flow.yml`. It is enabled by default. We agreed not to change `flow.yml`. Thank you.

**Comment by AmericanPresidentJimmyCarter** [comment]
The instructions in the README state to run `flow_parser.py` first. If I comment it out, `flow_parser.py` will not be able to generate the correct set of instructions in `flow.tmp.yml` for stable diffusion. In the `Dockerfile`, `start.sh` is used to start, which first calls `flow_parser.py`. I can also update the README for the native version to just say to use `start.sh`.

I can either duplicate the current `flow.yml` into something like `flow_with_sd.yml` and use it with the parser, but as the parser will already produce a stable-diffusion-less build in `flow.tmp.yml` that seems redundant.

**Comment by AmericanPresidentJimmyCarter** [comment]
Also please note that the default behavior of `flow_parser.yml` is that it generates a `flow.tmp.yml` with dalle-mega and glid3xl enabled but stable-diffusion disabled (see README).

**Comment by AmericanPresidentJimmyCarter** [comment]
Remaining k-samplers were drop in and seem to test OK, so I added them in. This should be ready for merge, let me know if there is anything else you might need.

File: README.md [hunk 6]
```diff
@@ -87,7 +88,7 @@ da = doc.matches
 da.plot_image_sprites(fig_size=(10,10), show_index=True)
 ```
 
-Here we generate 16 candidates, 8 from DALLE-mega and 8 from GLID3 XL, this is as defined in `num_images`, which takes about ~2 minutes. You can use a smaller value if it is too long for you. 
+Here we generate 24 candidates, 8 from DALLE-mega, 8 from GLID3 XL, and 8 from Stable Diffusion, this is as defined in `num_images`, which takes about ~2 minutes. You can use a smaller value if it is too long for you.
```
Comment (samsja) [inline README.md:91]:
```suggestion
Here we generate 16 candidates, 8 from DALLE-mega, 8 from GLID3 XL, and 8 from Stable Diffusion, this is as defined in `num_images`, which takes about ~2 minutes. You can use a smaller value if it is too long for you. 
```

File: README.md [hunk 7]
```diff
@@ -190,6 +192,20 @@ If you have installed Jina, the above flowchart can be generated via:
 jina export flowchart flow.yml flow.svg
 ```
 
+
+### Stable Diffusion weights
+
+If you want to use Stable Diffusion, you will first need to register an account on the website [Huggingface](https://huggingface.co/) and agree to the terms and conditions for the model. You can find the version of the model required by going here:
```
Comment (delgermurun) [inline README.md:198]:
```suggestion
If you want to use Stable Diffusion, you will first need to register an account on the website [Huggingface](https://huggingface.co/), go to the [repo](https://huggingface.co/CompVis/stable-diffusion-v-1-4-original), and agree to the terms and conditions for the model. You can find the version of the model required by going here:
```

If I go to directly download page without accepting terms, it show `403` page.

**Comment by delgermurun** [comment]
@AmericanPresidentJimmyCarter 

I can't stop docker process by clicking `ctrl+c`. I can only stop it by `docker stop <id>`.
Does it happen to you?

**Comment by AmericanPresidentJimmyCarter** [comment]
Yes, I figured that was normal but maybe it is not. I guess it's because the bash script starts `jina` rather than directly. I think `-it` just needs to be added to the `docker run` command, I can add that into the README.

File: README.md (hunk 7 above)
Comment (AmericanPresidentJimmyCarter) [reply README.md:198]: I will update the README with clearer instructions.

**Comment by AmericanPresidentJimmyCarter** [comment]
@delgermurun I have updated the README. The link for stable diffusion weights now works even if not logged in and prompts the user to login. T added `-it` to all example docker run commands and also formatted them multi-line to make them easier to read, since there were so many arguments.
